File: core-fs/src/main/java/com/ustadmobile/core/impl/HttpCache.java
package com.ustadmobile.core.impl;

import com.ustadmobile.core.controller.CatalogPresenter;
import com.ustadmobile.core.impl.http.UmHttpCall;
import com.ustadmobile.core.impl.http.UmHttpRequest;
import com.ustadmobile.core.impl.http.UmHttpResponse;
import com.ustadmobile.core.impl.http.UmHttpResponseCallback;
import com.ustadmobile.core.util.UMFileUtil;
import com.ustadmobile.core.util.UMIOUtils;
import com.ustadmobile.core.util.UMUUID;
import com.ustadmobile.core.util.UMUtil;

import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Hashtable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Created by mike on 12/26/17.
 *
 * The HttpCache provides transparent async http caching, as well as the ability to 'prime' the cache
 * by pre-loading it. It will be expanded to support saving cache entries in multiple directories so
 * that certain items can be 'subscribed' to and thus will not be eligible for deletion.
 *
 * The HttpCache will store etag and last modified information, and (mostly) obey http cache-control
 * and related headers.
 *
 * Placed in com.ustadmobile.core.impl so that the systemimpl makeRequestSync method can be
 * protected and accessed by this class, but it won't be accessible to other classes that
 * shouldn't use it (e.g. presenters).
 */
public class HttpCache {

    private String sharedDir;

    private String basePrivateDir;

    private JSONObject cacheDb;

    private final ExecutorService executorService = Executors.newFixedThreadPool(4);

    private int defaultTimeToLive = 60* 1000;

    public static final String PROTOCOL_FILE = "file://";

    /**
     * Wrapper HttpCall object that can be used to cancel the request if required. It implements
     * runnable and can be invoked using the executor service.
     */
    private class UmHttpCacheCall extends UmHttpCall implements Runnable{

        protected UmHttpRequest request;

        //Actual http request, if an outgoing http request is actually sent
        private UmHttpRequest httpRequest;

        private UmHttpResponseCallback responseCallback;//End response callback

        //Handler that deals with the http request, if an outgoing request is actually sent
        private UmHttpResponseHandler httpResponseHandler;

        protected Hashtable requestCacheControl;

        protected HttpCacheEntry entry;

        private UmHttpCacheCall(UmHttpRequest request, UmHttpResponseCallback responseCallback) {
            this.request = request;
            this.responseCallback = responseCallback;
        }


        @Override
        public void run() {
            final UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();

            if(request.getUrl().startsWith(PROTOCOL_FILE)) {
                File responseFile = new File(UMFileUtil.stripPrefixIfPresent("file://",
                        request.getUrl()));
                responseCallback.onComplete(this, new FileProtocolCacheResponse(responseFile));
                return;
            }

            HttpCacheEntry entry = getEntry(request.getUrl());
            if(entry != null) {
                long expirationTime = calculateExpiryTime(entry.getCacheControl(), entry.getExpiresTime(),
                        entry.getLastChecked() + defaultTimeToLive);
                if(request.isOnlyIfCached() ||
                        !(request.mustRevalidate() && expirationTime > System.currentTimeMillis())) {
                    HttpCacheResponse cacheResponse = new HttpCacheResponse(entry);
                    cacheResponse.setCacheResponse(HttpCacheResponse.HIT_DIRECT);
                    //no validation required - directly return the cached response
                    responseCallback.onComplete(this, cacheResponse);
                    return;
                }
            }else if(request.isOnlyIfCached() && entry == null) {
                responseCallback.onFailure(this, new FileNotFoundException(request.getUrl()));
                return;
            }

            //make an http request for this cache entry
            httpRequest = new UmHttpRequest(request.getUrl());
            if(entry != null) {
                if(entry.geteTag() != null) {
                    httpRequest.addHeader("if-none-match", entry.geteTag());
                }
                if(entry.getLastModified() > 0) {
                    httpRequest.addHeader("if-modified-since",
                            UMUtil.makeHTTPDate(entry.getLastModified()));
                }
            }

            httpResponseHandler = new UmHttpResponseHandler(this);
            impl.sendRequestAsync(httpRequest, httpResponseHandler);
        }


        @Override
        public void cancel() {
            //TODO: implement cancel
        }
    }

    /**
     * The HttpResponseHandler handles the response from the network after UmHttpCacheCall makes
     * an http call (if needed).
     */
    private class UmHttpResponseHandler implements Runnable, UmHttpResponseCallback {

        private UmHttpCacheCall cacheCall;

        private UmHttpResponse response;

        private UmHttpResponseHandler(UmHttpCacheCall cacheCall) {
            this.cacheCall = cacheCall;
        }

        @Override
        public void onComplete(UmHttpCall call, UmHttpResponse response) {
            this.response = response;
            executorService.execute(this);
        }

        @Override
        public void onFailure(UmHttpCall call, IOException exception) {
            cacheCall.responseCallback.onFailure(cacheCall, exception);
        }

        @Override
        public void run() {
            HttpCacheResponse cacheResponse = cacheResponse(cacheCall.request, response);
            cacheCall.responseCallback.onComplete(cacheCall, cacheResponse);
        }
    }



    public HttpCache(String sharedDir) {
        this.sharedDir = sharedDir;
        initCache();
    }

    protected void initCache() {
        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        InputStream fileIndexIn = null;
        try {
            if(!impl.fileExists(sharedDir)) {
                impl.makeDirectoryRecursive(sharedDir);
            }

            String sharedIndexPath = getSharedIndexPath();

            if(impl.fileExists(sharedIndexPath)) {
                ByteArrayOutputStream bout = new ByteArrayOutputStream();
                fileIndexIn = impl.openFileInputStream(sharedIndexPath);
                UMIOUtils.readFully(fileIndexIn, bout);
                cacheDb = new JSONObject(new String(bout.toByteArray(), "UTF-8"));
            }else {
                cacheDb = new JSONObject();
            }
        }catch(IOException e) {
            UstadMobileSystemImpl.l(UMLog.CRITICAL, 4, sharedDir, e);
        }finally {
            UMIOUtils.closeInputStream(fileIndexIn);
        }
    }

    protected String getSharedIndexPath() {
        return UMFileUtil.joinPaths(new String[]{sharedDir, "cache_index.json"});
    }

    public UmHttpCall get(UmHttpRequest request, UmHttpResponseCallback callback) {
        UmHttpCacheCall cacheCall = new UmHttpCacheCall(request, callback);
        executorService.execute(cacheCall);
        return cacheCall;
    }


    private HttpCacheEntry getEntry(String url) {
        JSONObject entryObj = cacheDb.optJSONObject(url);
        HttpCacheEntry entry = null;
        if(entryObj != null) {
            entry = new HttpCacheEntry();
            entry.loadFromJson(entryObj);
        }

        return entry;
    }

    public HttpCacheResponse cacheResponse(UmHttpRequest request, UmHttpResponse response) {
        HttpCacheEntry entry = getEntry(request.getUrl());
        if(entry == null) {
            entry = new HttpCacheEntry();
            entry.setFileUri(generateCacheEntryFileName(request, response, sharedDir));
        }

        HttpCacheResponse cacheResponse = new HttpCacheResponse(entry);

        if(response.getStatus() != 304) {
            cacheResponse.setCacheResponse(HttpCacheResponse.MISS);
            FileOutputStream fout = null;
            InputStream responseIn = null;
            try {
                fout = new FileOutputStream(entry.getFileUri());
                responseIn = response.getResponseAsStream();
                UMIOUtils.readFully(responseIn, fout);
            }catch(IOException e){
                UstadMobileSystemImpl.l(UMLog.ERROR, 66, null, e);
            }finally {
                UMIOUtils.closeInputStream(responseIn);
                UMIOUtils.closeOutputStream(fout);
            }
        }else {
            cacheResponse.setCacheResponse(HttpCacheResponse.HIT_VALIDATED);
        }

        long currentTime = System.currentTimeMillis();
        entry.updateFromResponse(response);
        entry.setLastAccessed(currentTime);
        entry.setLastChecked(currentTime);

        cacheDb.put(request.getUrl(), entry.toJson());

        return cacheResponse;
    }

    private String generateCacheEntryFileName(UmHttpRequest request, UmHttpResponse response,
                                              String dir) {
        File dirFile = new File(dir);
        File entryFile;
        String filename = CatalogPresenter.sanitizeIDForFilename(
                UMFileUtil.getFilename(request.getUrl()));

        String[] filenameParts = UMFileUtil.splitFilename(filename);
        String contentType = response.getHeader(UmHttpRequest.HEADER_CONTENT_TYPE);
        if(contentType != null) {
            String expectedExtension = UstadMobileSystemImpl.getInstance().getExtensionFromMimeType(contentType);
            if(expectedExtension != null && !expectedExtension.equals(filenameParts[1])) {
                filenameParts = new String[]{filename, expectedExtension};
                filename = filenameParts[0] + "." + filenameParts[1];
            }
        }

        entryFile = new File(dirFile, filename);
        if(!entryFile.exists()) {
            return entryFile.getAbsolutePath();
        }

        //try and get to a unique suffix
        for(int i = 0; i < 100; i++) {
            entryFile = new File(dir, filenameParts[0] + i + '.' + filenameParts[1]);
            if(!entryFile.exists())
                return entryFile.getAbsolutePath();
        }

        return UMUUID.randomUUID().toString() + "." + filenameParts[1];
    }

    /**
     * Calculates when an entry will expire based on it's HTTP headers: specifically
     * the expires header and cache-control header
     *
     * As per :  http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html section
     * 14.9.3 the max-age if present will take precedence over the expires header
     *
     *
     * @param cacheControlHeader Cache control header value
     * @param expires value (in utime) of the expires header
     *
     * @param defaultVal Expiry value to use in case headers do not contain max-age or expires
     *
     * @return
     */
    public static long calculateExpiryTime(String cacheControlHeader, long expires, long defaultVal) {
        if(cacheControlHeader != null) {
            Hashtable ccParams = UMFileUtil.parseParams(cacheControlHeader, ',');
            if(ccParams.containsKey("max-age")) {
                long maxage = Integer.parseInt((String)ccParams.get("max-age"));
                return System.currentTimeMillis() + (maxage * 1000);
            }
        }

        if(expires >= 0) {
            return expires;
        }

        return defaultVal;
    }


}

File: lib-core-fs/src/main/java/com/ustadmobile/core/impl/HttpCache.kt
package com.ustadmobile.core.impl

import com.ustadmobile.core.db.UmAppDatabase
import com.ustadmobile.core.impl.http.UmHttpCall
import com.ustadmobile.core.impl.http.UmHttpRequest
import com.ustadmobile.core.impl.http.UmHttpResponse
import com.ustadmobile.core.impl.http.UmHttpResponseCallback
import com.ustadmobile.core.util.UMCalendarUtil
import com.ustadmobile.core.util.UMFileUtil
import com.ustadmobile.core.util.UMIOUtils
import com.ustadmobile.core.util.UMUUID
import com.ustadmobile.lib.db.entities.HttpCachedEntry

import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import java.util.ArrayList
import java.util.Arrays
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Created by mike on 12/26/17.
 *
 * The HttpCache provides transparent synchronous and async http caching, as well as the ability to
 * 'prime' the cache by pre-loading it. It will be expanded to support saving cache entries in multiple
 * directories so that certain items can be 'subscribed' to and thus will not be eligible for deletion.
 *
 * The HttpCache will store etag and last modified information, and (mostly) obey http cache-control
 * and related headers.
 *
 * Placed in com.ustadmobile.core.impl so that the systemimpl makeRequestSync method can be
 * protected and accessed by this class, but it won't be accessible to other classes that
 * shouldn't use it (e.g. presenters).
 *
 * TODO: Answer HTTP head request with information from a cached GET request
 */
class HttpCache(private val sharedDir: String) : HttpCacheResponse.ResponseCompleteListener {

    private val basePrivateDir: String? = null

    private val executorService = Executors.newFixedThreadPool(4)

    private val defaultTimeToLive = DEFAULT_TIME_TO_LIVE

    /**
     * Wrapper HttpCall object that can be used to cancel the request if required. It implements
     * runnable and can be invoked using the executor service.
     */
    private inner class UmHttpCacheCall private constructor(var request: UmHttpRequest, private val responseCallback: UmHttpResponseCallback?//End response callback
    ) : UmHttpCall(), Runnable {

        //Actual http request, if an outgoing http request is actually sent
        private var httpRequest: UmHttpRequest? = null

        //Handler that deals with the http request, if an outgoing request is actually sent
        private var httpResponseHandler: UmHttpResponseHandler? = null

        protected var entry: HttpCachedEntry? = null

        private var async = true

        init {

            async = responseCallback != null
        }

        @Throws(IOException::class)
        fun execute(): UmHttpResponse? {
            val impl = UstadMobileSystemImpl.instance
            var cacheResponse: AbstractCacheResponse? = null

            if (request.url!!.startsWith(PROTOCOL_FILE)) {
                val filePath = UMFileUtil.stripPrefixIfPresent("file://", request.url!!)
                val zipSepPos = filePath.indexOf('!')
                if (zipSepPos == -1) {
                    val responseFile = File(filePath)
                    cacheResponse = FileProtocolCacheResponse(responseFile)
                    if (async) {
                        responseCallback!!.onComplete(this, cacheResponse)
                    }
                } else {
                    cacheResponse = ZipEntryCacheResponse(
                            File(filePath.substring(0, zipSepPos)),
                            filePath.substring(zipSepPos + 1))
                    if (async) {
                        responseCallback!!.onComplete(this, cacheResponse)
                    }
                }

                return cacheResponse
            }

            val entry = getEntry(request.context, request.url, request.getMethod())
            if (entry != null) {
                val timeToLive = if (request.mustRevalidate()) 0 else defaultTimeToLive
                if (isFresh(entry, timeToLive) || request.isOnlyIfCached()) {
                    cacheResponse = HttpCacheResponse(entry, request)
                    cacheResponse.cacheResponse = HttpCacheResponse.HIT_DIRECT
                    UstadMobileSystemImpl.l(UMLog.ERROR, 384, "Cache:HIT_DIRECT: " + request.url!!)

                    //no validation required - directly return the cached response
                    if (async) {
                        responseCallback!!.onComplete(this, cacheResponse)
                    }
                    return cacheResponse

                }
            } else if (request.isOnlyIfCached() && entry == null) {
                val ioe = FileNotFoundException(request.url)
                if (async) {
                    UstadMobileSystemImpl.l(UMLog.INFO, 386,
                            "Cache:onlyIfCached: Fail: not cached: " + request.url!!)
                    responseCallback!!.onFailure(this, ioe)
                    return cacheResponse
                } else {
                    throw ioe
                }
            }

            //make an http request for this cache entry
            httpRequest = UmHttpRequest(request.context!!, request.url!!)
            if (entry != null) {
                if (entry.etag != null) {
                    httpRequest!!.addHeader("if-none-match", entry.etag)
                }
                if (entry.lastModified > 0) {
                    httpRequest!!.addHeader("if-modified-since",
                            UMCalendarUtil.makeHTTPDate(entry.lastModified))
                }
            }

            httpResponseHandler = UmHttpResponseHandler(this)
            if (async) {
                impl.sendRequestAsync(httpRequest!!, httpResponseHandler!!)
            } else {
                httpResponseHandler!!.response = impl.sendRequestSync(httpRequest!!)
                return httpResponseHandler!!.execute()

            }

            return null
        }

        override fun run() {
            try {
                execute()
            } catch (e: IOException) {
                UstadMobileSystemImpl.l(UMLog.ERROR, 73, request.url, e)
            }

        }


        override fun cancel() {
            //TODO: implement cancel
        }
    }

    /**
     * The HttpResponseHandler handles the response from the network after UmHttpCacheCall makes
     * an http call (if needed).
     */
    private inner class UmHttpResponseHandler private constructor(private val cacheCall: UmHttpCacheCall) : Runnable, UmHttpResponseCallback {

        private var response: UmHttpResponse? = null

        override fun onComplete(call: UmHttpCall, response: UmHttpResponse) {
            this.response = response
            executorService.execute(this)
        }

        override fun onFailure(call: UmHttpCall, exception: IOException) {
            cacheCall.responseCallback!!.onFailure(cacheCall, exception)
        }

        override fun run() {
            execute()
        }

        fun execute(): AbstractCacheResponse {
            val responseCacheControlHeader = response!!.getHeader(UmHttpRequest.HEADER_CACHE_CONTROL)

            if (responseCacheControlHeader != null) {
                val responseCacheControl = UMFileUtil.parseParams(responseCacheControlHeader, ',')

                if (responseCacheControl.containsKey(CACHE_CONTROL_NO_CACHE)) {
                    val noCacheResponse = NoCacheResponse(response)
                    if (cacheCall.async) {
                        cacheCall.responseCallback!!.onComplete(cacheCall, noCacheResponse)
                    }
                    return noCacheResponse
                }
            }

            val cacheResponse = cacheResponse(cacheCall.request, response, !cacheCall.async)

            if (cacheCall.async) {
                cacheCall.responseCallback!!.onComplete(cacheCall, cacheResponse)
            }

            return cacheResponse
        }
    }

    private inner class DeleteEntriesTask private constructor(private val context: Any, private val urlsToDelete: Array<String>, private val callback: UmCallback<*>?) : Runnable {

        override fun run() {
            val fileUrisToDelete = UmAppDatabase.getInstance(context).httpCachedEntryDao
                    .findFileUrisByUrl(Arrays.asList(*urlsToDelete))
            val deletedFileUris = ArrayList<String>()
            var entryFile: File
            for (fileUri in fileUrisToDelete) {
                entryFile = File(fileUri)
                if (!entryFile.exists() || entryFile.exists() && entryFile.delete()) {
                    deletedFileUris.add(fileUri)
                } else {
                    UstadMobileSystemImpl.l(UMLog.ERROR, 0, "Failed to deleteByDownloadSetUid cache file: $fileUri")
                }
            }

            UmAppDatabase.getInstance(context).httpCachedEntryDao.deleteByFileUris(deletedFileUris)

            callback?.onSuccess(null)
        }
    }


    init {
        initCache()
    }

    protected fun initCache() {
        val impl = UstadMobileSystemImpl.instance
        val fileIndexIn: InputStream? = null
        //        This class is going to be removed anyway and replaced with using the image library caching mechanisms
        //        try {
        //            if(!new File(sharedDir).exists()) {
        //                impl.makeDirectoryRecursive(sharedDir);
        //            }
        //        }catch(IOException e) {
        //            UstadMobileSystemImpl.l(UMLog.CRITICAL, 4, sharedDir, e);
        //        }finally {
        //            UMIOUtils.closeInputStream(fileIndexIn);
        //        }
    }

    operator fun get(request: UmHttpRequest, callback: UmHttpResponseCallback): UmHttpCall {
        val cacheCall = UmHttpCacheCall(request, callback)
        executorService.execute(cacheCall)
        return cacheCall
    }

    /**
     * Performs an http request synchronously. Returns as soon as the the connection is established.
     * The response will be simultaneously saved to the disk as it's read. If the response comes from
     * the network, this is done using an executor to fork a thread that simultaneously writes to a
     * fileoutputstream to cache the entry to disk and writes to a pipedoutputstream, which connects
     * with a pipedinputstream providing the response to the consumer.
     *
     * The client is expected to read the entire response. Failing to do so could cause an issue as
     * the pipedstream has a limited buffer size.
     *
     * @param request HttpRequest object for the request
     *
     * @return AbstractCacheResponse object representing the http response.
     * @throws IOException If an IOException occurs
     */
    @Throws(IOException::class)
    fun getSync(request: UmHttpRequest): UmHttpResponse? {
        val cacheCall = UmHttpCacheCall(request, null)
        return cacheCall.execute()
    }

    private fun getEntry(context: Any?, url: String?, method: String): HttpCachedEntry? {
        return UmAppDatabase.getInstance(context).httpCachedEntryDao
                .findByUrlAndMethod(url, getMethodFlag(method))
    }

    fun cacheResponse(request: UmHttpRequest, networkResponse: UmHttpResponse,
                      forkSaveToDisk: Boolean): HttpCacheResponse {
        val requestUrl = request.url
        var entry = getEntry(request.context, requestUrl, request.getMethod())
        val responseHasBody = UmHttpRequest.METHOD_HEAD != request.getMethod() && networkResponse.status != 204

        if (entry == null) {
            entry = HttpCachedEntry()
            entry.url = requestUrl
            if (responseHasBody) {
                entry.fileUri = generateCacheEntryFileName(request, networkResponse, sharedDir)
            }
        }

        val cacheResponse = HttpCacheResponse(entry, request)
        cacheResponse.networkResponse = networkResponse
        cacheResponse.entry.lastChecked = System.currentTimeMillis()
        updateCachedEntryFromNetworkResponse(cacheResponse.entry, networkResponse)


        if (networkResponse.status == 304) {
            updateCacheIndex(cacheResponse)
            cacheResponse.isNetworkResponseNotModified = true
            UstadMobileSystemImpl.l(UMLog.INFO, 387, "Cache:HIT_VALIDATED:" + request.url!!)
        } else if (responseHasBody) {
            cacheResponse.setOnResponseCompleteListener(this)
            UstadMobileSystemImpl.l(UMLog.INFO, 385, "Cache:MISS - storing:" + request.url!!)
            if (forkSaveToDisk) {
                cacheResponse.initPipe()
                executorService.execute(cacheResponse)
            } else {
                cacheResponse.saveNetworkResponseToDiskAndBuffer()
            }
        }

        return cacheResponse
    }


    fun deleteEntries(context: Any, urls: Array<String>, callback: UmCallback<*>) {
        executorService.execute(DeleteEntriesTask(context, urls, callback))
    }

    fun deleteEntriesSync(context: Any, urls: Array<String>) {
        DeleteEntriesTask(context, urls, null).run()
    }

    protected fun updateCacheIndex(response: HttpCacheResponse) {
        if (response.networkResponse!!.status == 304) {
            response.cacheResponse = HttpCacheResponse.HIT_VALIDATED
        }

        response.entry.lastAccessed = System.currentTimeMillis()
        UmAppDatabase.getInstance(response.request.context).httpCachedEntryDao
                .insert(response.entry)
    }

    override fun onResponseComplete(response: HttpCacheResponse) {
        updateCacheIndex(response)
    }


    private fun generateCacheEntryFileName(request: UmHttpRequest, response: UmHttpResponse,
                                           dir: String): String {
        val dirFile = File(dir)
        var entryFile: File
        var filename = UMIOUtils.sanitizeIDForFilename(
                UMFileUtil.getFilename(request.url!!))

        var filenameParts = UMFileUtil.splitFilename(filename)
        val contentType = response.getHeader(UmHttpRequest.HEADER_CONTENT_TYPE)
        if (contentType != null) {
            val expectedExtension = UstadMobileSystemImpl.instance.getExtensionFromMimeType(contentType)
            if (expectedExtension != null && expectedExtension != filenameParts[1]) {
                filenameParts = arrayOf(filename, expectedExtension)
                filename = filenameParts[0] + "." + filenameParts[1]
            }
        }

        entryFile = File(dirFile, filename)
        if (!entryFile.exists()) {
            return entryFile.absolutePath
        }

        //try and get to a unique suffix
        for (i in 0..99) {
            entryFile = File(dir, filenameParts[0] + i + '.'.toString() + filenameParts[1])
            if (!entryFile.exists())
                return entryFile.absolutePath
        }

        return File(dir, UMUUID.randomUUID().toString() + "." + filenameParts[1]).absolutePath
    }

    companion object {

        val DEFAULT_TIME_TO_LIVE = 60 * 60 * 1000

        val PROTOCOL_FILE = "file://"

        val CACHE_CONTROL_KEY_MAX_AGE = "max-age"

        val CACHE_CONTROL_NO_CACHE = "no-cache"

        /**
         * Update the given HttpCachedEntry from the response received over the network. This can't be
         * part of the entry  itself, because the entry is in the database module. This will only updateState
         * the object itself and will NOT persist it to the database
         *
         * @param cachedEntry The cached entry to updateState
         * @param networkResponse The network response just received from the network
         */
        fun updateCachedEntryFromNetworkResponse(cachedEntry: HttpCachedEntry, networkResponse: UmHttpResponse) {
            val headerVal: String?
            if (networkResponse.status != 304) {
                //new entry was downloaded - updateState the length etc.
                headerVal = networkResponse.getHeader(UmHttpRequest.HEADER_CONTENT_LENGTH)
                if (headerVal != null) {
                    try {
                        cachedEntry.contentLength = Integer.parseInt(headerVal).toLong()
                    } catch (e: IllegalArgumentException) {
                        UstadMobileSystemImpl.l(UMLog.ERROR, 74, headerVal, e)
                    }

                }

                cachedEntry.statusCode = networkResponse.status
            }

            cachedEntry.cacheControl = networkResponse.getHeader(UmHttpRequest.HEADER_CACHE_CONTROL)
            cachedEntry.contentType = networkResponse.getHeader(UmHttpRequest.HEADER_CONTENT_TYPE)
            cachedEntry.expiresTime = convertDateHeaderToLong(UmHttpRequest.HEADER_EXPIRES, networkResponse)
            cachedEntry.contentType = networkResponse.getHeader(UmHttpRequest.HEADER_CONTENT_TYPE)
            cachedEntry.etag = networkResponse.getHeader(UmHttpRequest.HEADER_ETAG)
            cachedEntry.cacheControl = networkResponse.getHeader(UmHttpRequest.HEADER_CACHE_CONTROL)
        }

        /**
         * Determine if the entry is considered fresh.
         *
         * @see .calculateEntryExpirationTime
         * @param timeToLive the time since when this entry was last checked for which the entry will be
         * considered fresh if the cache-control headers and expires headers do not
         * provide this information.
         *
         * @return true if the entry is considered fresh, false otherwise
         */
        @JvmOverloads
        fun isFresh(cachedEntry: HttpCachedEntry, timeToLive: Int = DEFAULT_TIME_TO_LIVE): Boolean {
            val expiryTime = calculateEntryExpirationTime(cachedEntry)
            val timeNow = System.currentTimeMillis()

            return if (expiryTime != -1) {
                expiryTime > timeNow
            } else {
                cachedEntry.lastChecked + timeToLive > timeNow
            }
        }

        /**
         * Calculates when an entry will expire based on it's HTTP headers: specifically
         * the expires header and cache-control header
         *
         * As per :  http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html section
         * 14.9.3 the max-age if present will take precedence over the expires header
         *
         * @return -1 if the expiration time calculated from the headers provided if possible, -1 otherwise
         */
        fun calculateEntryExpirationTime(cachedEntry: HttpCachedEntry): Long {
            val cacheControl = cachedEntry.cacheControl
            if (cacheControl != null) {
                val ccParams = UMFileUtil.parseParams(cacheControl, ',')
                if (ccParams.containsKey(CACHE_CONTROL_KEY_MAX_AGE)) {
                    val maxage = Integer.parseInt(ccParams[CACHE_CONTROL_KEY_MAX_AGE] as String).toLong()
                    return cachedEntry.lastChecked + maxage * 1000
                }
            }

            return if (cachedEntry.expiresTime >= 0) {
                cachedEntry.expiresTime
            } else -1

        }


        private fun convertDateHeaderToLong(headerName: String, response: UmHttpResponse): Long {
            val headerVal = response.getHeader(headerName)
            return if (headerVal != null) {
                try {
                    UMCalendarUtil.parseHTTPDate(headerVal)
                } catch (e: NumberFormatException) {
                    -1L
                }

            } else {
                -1L
            }
        }

        /**
         * Get the METHOD_ integer flag for the given HTTP Method as a string
         *
         * @param methodName The HTTP method as a string e.g. "GET", "HEAD", "POST"
         *
         * @return
         */
        fun getMethodFlag(methodName: String): Int {
            if (methodName.equals(UmHttpRequest.METHOD_GET, ignoreCase = true))
                return HttpCachedEntry.METHOD_GET
            else if (methodName.equals(UmHttpRequest.METHOD_HEAD, ignoreCase = true))
                return HttpCachedEntry.METHOD_HEAD
            else if (methodName.equals(UmHttpRequest.METHOD_POST, ignoreCase = true))
                return HttpCachedEntry.METHOD_POST

            return -1
        }
    }

}
